### ISSUE TITLE
Extern fn declarations can have dangling lifetimes

### ISSUE THREAD

**Issue by Manishearth** [rust-lang/rust#35851]
``` rust
extern {
    fn foo() -> &u8;
}
```

([playpen](https://is.gd/9PaZfM))

compiles fine. The `&u8` has a dangling lifetime, which doesn't compile for functions that have bodies.

This probably should throw an error.

**Comment by Manishearth** [comment]
Probably would be a breaking change to fix. But soundness.

**Comment by eddyb** [comment]
Elision was never added to them. cc @rust-lang/lang

**Comment by nagisa** [comment]
Seems fine to break to me. References in extern blocks is very likely not what you want anyway (i.e. is a bug).

**Comment by Manishearth** [comment]
Well, it also happens with lifetime parameter elision. For the servo-gecko stuff we've managed to configure bindgen to return typedefs to types with lifetimes for things which are conceptually borrows. So we're using this _today_ :smile:

(thats how I discovered this, someone else added a dangling lifetime and I was bewildered by it. We're going to merge it for now because my safer bindings glue isn't complete yet, so the input lifetime for that function can't exist yet, but it will soon)

**Comment by nikomatsakis** [comment]
Discussed in @rust-lang/lang meeting. This is definitely a bug; we should "just fix it". We should measure the impact, of course.

**Comment by nikomatsakis** [comment]
triage: P-medium

**Comment by eddyb** [comment]
I accidentally unified the two codepaths in my `lazy-static` so consider the impact measured, I guess?
* [ ] [`liquidfun`](https://taskcluster.github.io/unified-logviewer/?url=https%3A%2F%2Fqueue.taskcluster.net%2Fv1%2Ftask%2Fg2mnllcXRrmF1iO1W34s1Q%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log&highlightStart=92&highlightEnd=92&lineNumber=&wrapLines=false&showLineNumbers=true&jumpToHighlight=true)
* [ ] [`xcb`](https://tools.taskcluster.net/task-inspector/#7_lGM7ABTOiDJqb8D5YgtQ/) ([logs](https://taskcluster.github.io/unified-logviewer/?url=https%3A%2F%2Fqueue.taskcluster.net%2Fv1%2Ftask%2F7_lGM7ABTOiDJqb8D5YgtQ%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log&highlightStart=91&highlightEnd=91&lineNumber=&wrapLines=false&showLineNumbers=true&jumpToHighlight=true)

**Comment by Mark-Simulacrum** [comment]
@eddyb So I presume that we don't currently have the lazy-start branch in tree, since this hasn't been fixed. Perhaps the two codepaths were un-unified?

**Comment by eddyb** [comment]
Yeah, I redid some things, and tried to avoid regressing *anything at all*.

**Comment by nikomatsakis** [comment]
It would probably be good to deprecate (future compat warning?) code using this pattern.

**Comment by nikomatsakis** [comment]
Well, for sure we should deprecate. It doesn't necessarily *have* to be a future-compat warning. i.e., we could live with this.

**Comment by Centril** [comment]
@eddyb @nagisa The snippet in the OP does not compile today... is there anything left to do here?

**Comment by kenta7777** [comment]
I'd like to tackle this issue. How shall we begin?

**Comment by iluuu1994** [comment]
Seems fixed, there's a test for it [here](https://github.com/rust-lang/rust/blob/master/src/test/ui/foreign-fn-return-lifetime.rs).

**Comment by Centril** [comment]
Indeed it seems so; closing therefore.
